### PR TITLE
Kokkos + KokkosKernels Promotion To 4.1.00

### PR DESCRIPTION
**Pull request by ndellingwood**
@trilinos/kokkos 
@trilinos/kokkos-kernels 
Adding @trilinos/tpetra to notify of update to `Tpetra_SUPPORTED_KOKKOS_VERSION`

## Motivation

Release Kokkos Ecosystem 4.1.00 into Trilinos

## Testing
Various nightly testing configurations and upcoming AT testing

### CODE REVIEW

**Comment by ndellingwood** [comment]
Cross-referencing stand-alone repo release PRs:
https://github.com/kokkos/kokkos/pull/6231
https://github.com/kokkos/kokkos-kernels/pull/1877

**Comment by ndellingwood** [comment]
Adding mentions to members of app teams interested in testing this branch and tracking this PR:

@rppawlo for EMPIRE
@stanmoore1 for LAMMPS
@spdomin for NALU
@vbrunini for ARIA
@mperego @jewatkins for Albany and/or SPARC

**Comment by jewatkins** [comment]
We're still trying to get kokkos 4.0 fully working :)

File: cmake/std/atdm/contributed/blake/environment.sh [hunk 1]
```diff
@@ -97,48 +106,47 @@ else
...
-  export LOCAL_LAPACK_LIBRARIES=${LOCAL_BLAS_LIBRARIES}
+if [[ "$ATDM_CONFIG_COMPILER" == "INTEL-2021.4.0" ]] ; then
+  module load intel/oneAPI/hpc-toolkit/2021.4.0 intel/oneAPI/base-toolkit/2021.4.0 
+  module load openmpi/4.0.5/intel-oneapi/2021.4.0 
+  module swap gcc/7.2.0 gcc/10.2.0
+  module load openblas/0.3.21/gcc/10.2.0
+  module load boost/1.75.0/intel-oneapi/2021.2.0
+  module load hdf5/1.10.7/openmpi/4.0.5/intel-oneapi/2021.2.0 netcdf-c/4.7.4/openmpi/4.0.5/intel-oneapi/2021.2.0 zlib/1.2.11
+  module load python/3.7.3
+  #module load python/2.7.13
```
Comment (jewatkins) [inline cmake/std/atdm/contributed/blake/environment.sh:117]: Has this been tested on blake? Might try to use this for Albany.

Comment (ndellingwood) [reply cmake/std/atdm/contributed/blake/environment.sh:117]: @jewatkins I've been running the gnu/11 option in a nightly build (OpenMP+Serial, debug build, SKX arch) using the [cmake/std/atdm/contributed/blake/environment.sh](https://github.com/trilinos/Trilinos/pull/11988/files/b36484ff1af4937557a3aa160afafea3af5ae849#diff-8e7cd8c90aa32e48f27bad72361a472bb53331c125743e66b3a413f718c89886) script

I've been testing with the intel/2021.4 changes but have run into issues with netcdf when trying to enable SEACAS and have not worked that out yet (not sure if module/system related or user error)

Comment (ndellingwood) [reply cmake/std/atdm/contributed/blake/environment.sh:117]: @jewatkins also, lots of failing unit tests in OpenMP builds with intel/2021.4 (various issues opened)

Comment (jewatkins) [reply cmake/std/atdm/contributed/blake/environment.sh:117]: Ah okay, we've run into similar issues. We decided to just drop our intel build and use the gcc build. I mentioned the issues with intel modules to admins but I haven't heard of anything new coming.

**Comment by ndellingwood** [comment]
Updated the Kokkos snapshot (changelog fixup)

**Comment by ndellingwood** [comment]
Adding @dalg24 @masterleinad for tracking

**Comment by ndellingwood** [comment]
> Is that just easier to maintain than moving everything over to GenConfig?

@bartlettroscoe it was a fast and easy step to update those scripts based some of the builds I had worked on with those machines, the follow up step will be to add ini files for GenConfig

**Comment by ndellingwood** [comment]
The common error in the `Trilinos_PR_gcc-8.3.0` build is due to `-Werror=unused-variable`:

```
In file included from ../Trilinos/packages/kokkos/core/src/Kokkos_Parallel.hpp:31,
                 from ../Trilinos/packages/kokkos/core/src/OpenMP/Kokkos_OpenMP.hpp:39,
                 from ../Trilinos/packages/kokkos/core/src/decl/Kokkos_Declare_OPENMP.hpp:21,
                 from packages/kokkos/KokkosCore_Config_DeclareBackend.hpp:22,
                 from ../Trilinos/packages/kokkos/core/src/Kokkos_Core.hpp:45,
                 from ../Trilinos/packages/tpetra/core/compat/Tpetra_KokkosCompat_ClassicNodeAPI_Wrapper.hpp:4,
                 from ../Trilinos/packages/tpetra/core/src/Tpetra_Details_DefaultTypes.hpp:45,
                 from ../Trilinos/packages/tpetra/core/src/Tpetra_CrsMatrix_fwd.hpp:45,
                 from ../Trilinos/packages/tpetra/core/src/Tpetra_CrsMatrix_decl.hpp:47,
                 from packages/tpetra/core/src/Tpetra_CrsMatrix_LONG_LONG_INT_LONG_LONG_OPENMP.cpp:66:
../Trilinos/packages/tpetra/core/src/Tpetra_Import_Util2.hpp: In instantiation of ‘void Tpetra::Import_Util::reverseNeighborDiscovery(const Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>&, const typename Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::row_ptrs_host_view_type&, const typename Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::local_inds_host_view_type&, const Tpetra::Details::Transfer<LocalOrdinal, GlobalOrdinal, Node>&, Teuchos::RCP<const Tpetra::Import<LocalOrdinal, GlobalOrdinal, Node> >, Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node> >, Teuchos::ArrayRCP<int>&, Teuchos::ArrayRCP<T2>&, Teuchos::RCP<const Teuchos::Comm<int> >&) [with Scalar = long long int; LocalOrdinal = int; GlobalOrdinal = long long int; Node = Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::OpenMP>; typename Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::row_ptrs_host_view_type = Kokkos::View<const long unsigned int*, Kokkos::LayoutRight, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::Experimental::EmptyViewHooks, Kokkos::MemoryTraits<0> >; typename Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::local_inds_host_view_type = Kokkos::View<const int*, Kokkos::LayoutRight, Kokkos::Device<Kokkos::OpenMP, Kokkos::HostSpace>, Kokkos::Experimental::EmptyViewHooks, Kokkos::MemoryTraits<0> >]’:
../Trilinos/packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp:8733:50:   required from ‘void Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::transferAndFillComplete(Teuchos::RCP<Tpetra::CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> >&, const Tpetra::Details::Transfer<LocalOrdinal, GlobalOrdinal, Node>&, const Teuchos::RCP<const Tpetra::Details::Transfer<LocalOrdinal, GlobalOrdinal, Node> >&, const Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node> >&, const Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node> >&, const Teuchos::RCP<Teuchos::ParameterList>&) const [with Scalar = long long int; LocalOrdinal = int; GlobalOrdinal = long long int; Node = Tpetra::KokkosCompat::KokkosDeviceWrapperNode<Kokkos::OpenMP>]’
packages/tpetra/core/src/Tpetra_CrsMatrix_LONG_LONG_INT_LONG_LONG_OPENMP.cpp:74:3:   required from here
../Trilinos/packages/kokkos/core/src/Kokkos_View.hpp:512:47: error: unused variable ‘Kokkos::AllowPadding’ [-Werror=unused-variable]
 inline constexpr Kokkos::Impl::AllowPadding_t AllowPadding{};
```

**Comment by bartlettroscoe** [comment]
> > Is that just easier to maintain than moving everything over to GenConfig?
> 
> @bartlettroscoe it was a fast and easy step to update those scripts based some of the builds I had worked on with those machines, the follow up step will be to add ini files for GenConfig

@ndellingwood, let us know how that goes and compare it to the work for the ATDM Trilinos configuration system, in your experience.

**Comment by ndellingwood** [comment]
Pushed commit [a5acbeb](https://github.com/trilinos/Trilinos/pull/11988/commits/a5acbeb097e9a70da5ea951c6a1f33c97287db28) to remove the unused using statements from Tpetra where the -Werror showed up in the Trilinos_PR_gcc-8.3.0 , will see how the next round of testing looks

**Review by crtrott — APPROVED** [review]

**Comment by rppawlo** [comment]
Good to go for EMPIRE!

**Pull request merged**